File: tooling/validate_skills.py
#!/usr/bin/env python3
"""Validate every skill under skills/ against the Agent Skills spec plus
this repository's own rules.

Spec rules (agentskills.io/specification, retrieved 2026-07-17):
  - name: 1-64 chars, lowercase a-z 0-9 and single hyphens, no leading/
    trailing hyphen, no consecutive hyphens, must equal the directory name
  - description: 1-1024 chars
  - compatibility: <=500 chars when present
  - metadata: mapping of string -> string when present

Repository rules:
  - SKILL.md body <=500 lines and under ~5000 estimated tokens
  - frontmatter metadata must carry the version contract keys
  - every relative path mentioned in SKILL.md exists
  - every file in references/ is mentioned in SKILL.md (the agent must be
    told when to load each reference)
  - no absolute filesystem paths anywhere in skill content

The official reference validator (`pip install skills-ref`, CLI
`agentskills validate`) runs in CI in addition to this script.

Usage: python3 tooling/validate_skills.py
"""

from __future__ import annotations

import re
import sys
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parent.parent
SKILLS_DIR = REPO_ROOT / "skills"

NAME_RE = re.compile(r"^[a-z0-9]+(-[a-z0-9]+)*$")

REQUIRED_METADATA_KEYS = {
    "skill-version",
    "verified-date",
    "verified-platform",
    "verified-targets",
}
# Per-skill component pins (moonbit-language pins moonc; moonbit-toolchain
# pins moon + moonrun). Checked here for presence; consistency with the
# toolchain snapshot is tooling/check_versions.py's job.
REQUIRED_COMPONENT_KEYS = {
    "moonbit-language": {"moonc-version"},
    "moonbit-toolchain": {"moon-version", "moonrun-version"},
}


def parse_frontmatter(text: str) -> tuple[dict, str, list[str]]:
    """Minimal YAML-subset frontmatter parser (scalars + one-level maps).

    Returns (frontmatter, body, errors). Deliberately strict: this repo's
    frontmatter must stay in the simple subset every client can parse.
    """
    errors: list[str] = []
    if not text.startswith("---\n"):
        return {}, text, ["missing frontmatter opening '---'"]
    try:
        end = text.index("\n---\n", 4)
    except ValueError:
        return {}, text, ["missing frontmatter closing '---'"]
    raw, body = text[4:end], text[end + 5 :]

    data: dict = {}
    current_map: dict | None = None
    for line_no, line in enumerate(raw.splitlines(), start=2):
        if not line.strip() or line.lstrip().startswith("#"):
            continue
        if line.startswith("  ") and current_map is not None:
            m = re.match(r"^  ([A-Za-z0-9_-]+):\s*(.*)$", line)
            if not m:
                errors.append(f"line {line_no}: unparseable nested entry {line!r}")
                continue
            current_map[m.group(1)] = m.group(2).strip().strip("\"'")
        else:
            m = re.match(r"^([A-Za-z0-9_-]+):\s*(.*)$", line)
            if not m:
                errors.append(f"line {line_no}: unparseable entry {line!r}")
                continue
            key, value = m.group(1), m.group(2).strip()
            if value == "":
                current_map = {}
                data[key] = current_map
            else:
                data[key] = value.strip().strip("\"'")
                current_map = None
    return data, body, errors


def validate_skill(skill_dir: Path) -> list[str]:
    problems: list[str] = []
    skill_md = skill_dir / "SKILL.md"
    if not skill_md.is_file():
        return [f"{skill_dir.name}: missing SKILL.md"]
    text = skill_md.read_text()
    fm, body, fm_errors = parse_frontmatter(text)
    problems += [f"frontmatter: {e}" for e in fm_errors]

    name = fm.get("name", "")
    if not isinstance(name, str) or not NAME_RE.match(name) or len(name) > 64:
        problems.append(f"name {name!r} violates spec naming rules")
    if name != skill_dir.name:
        problems.append(f"name {name!r} != directory name {skill_dir.name!r}")

    description = fm.get("description", "")
    if not isinstance(description, str) or not (1 <= len(description) <= 1024):
        problems.append(
            f"description length {len(description)} outside 1..1024 chars"
        )

    compatibility = fm.get("compatibility", "")
    if compatibility and len(compatibility) > 500:
        problems.append(f"compatibility length {len(compatibility)} > 500 chars")

    metadata = fm.get("metadata", {})
    if not isinstance(metadata, dict):
        problems.append("metadata must be a mapping")
        metadata = {}
    for key, value in metadata.items():
        if not isinstance(value, str):
            problems.append(f"metadata.{key} must be a string")
    required = REQUIRED_METADATA_KEYS | REQUIRED_COMPONENT_KEYS.get(name, set())
    missing = required - set(metadata)
    if missing:
        problems.append(f"metadata missing version-contract keys: {sorted(missing)}")

    body_lines = body.count("\n") + 1
    if body_lines > 500:
        problems.append(f"SKILL.md body has {body_lines} lines (> 500)")
    est_tokens = len(body) // 4
    if est_tokens > 5000:
        problems.append(f"SKILL.md body ~{est_tokens} tokens (> 5000)")

    # Relative path references in SKILL.md must exist.
    referenced: set[str] = set()
    for match in re.finditer(
        r"(?<![A-Za-z0-9_./-])((?:references|scripts|assets)/[A-Za-z0-9_./-]+[A-Za-z0-9_])",
        body,
    ):
        referenced.add(match.group(1))
    for rel in sorted(referenced):
        if not (skill_dir / rel).is_file():
            problems.append(f"SKILL.md references missing file {rel}")

    # Every reference/script file must be mentioned so agents know when to
    # load it (progressive disclosure requires an explicit routing hint).
    for sub in ("references", "scripts"):
        subdir = skill_dir / sub
        if not subdir.is_dir():
            continue
        for f in sorted(subdir.iterdir()):
            if f.name.startswith(".") or not f.is_file():
                continue
            rel = f"{sub}/{f.name}"
            # A generated .md twin of an .mbt.md source counts via its source.
            if rel not in referenced and rel not in body:
                problems.append(f"{rel} exists but SKILL.md never mentions it")

    for f in skill_dir.rglob("*"):
        if f.is_file() and f.suffix in (".md", ".mbt", ".sh", ".py"):
            if re.search(r"(?<![\w@])/(?:Users|home|private/tmp)/", f.read_text()):
                problems.append(f"{f.relative_to(skill_dir)}: absolute path leaked")

    return [f"{skill_dir.name}: {p}" for p in problems]


def main() -> int:
    skill_dirs = sorted(d for d in SKILLS_DIR.iterdir() if d.is_dir())
    if not skill_dirs:
        print("no skills found", file=sys.stderr)
        return 2
    all_problems: list[str] = []
    for skill_dir in skill_dirs:
        all_problems += validate_skill(skill_dir)
    for problem in all_problems:
        print(f"FAIL {problem}", file=sys.stderr)
    if not all_problems:
        print(f"validated {len(skill_dirs)} skill(s): OK")
    return 1 if all_problems else 0


if __name__ == "__main__":
    sys.exit(main())

File: tooling/validate_skills.ts
import { existsSync, readdirSync, readFileSync, statSync } from "node:fs";
import { basename, extname, join, relative } from "node:path";
import { exitWith, isMain } from "./lib/cli.ts";
import { parseFrontmatter, scalar, stringMap } from "./lib/frontmatter.ts";
import { REPO_ROOT } from "./lib/repo.ts";

const SKILLS_DIR = join(REPO_ROOT, "skills");

export const NAME_RE = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;

const REQUIRED_METADATA_KEYS = new Set(["skill-version", "updated-date"]);

export const SKILL_VERSION_RE = /^(?:0|[1-9]\d*)\.(?:0|[1-9]\d*)\.(?:0|[1-9]\d*)$/;
const DATE_RE = /^\d{4}-\d{2}-\d{2}$/;

const REQUIRED_COMPONENT_KEYS: Record<string, Set<string>> = {
  "moonbit-language": new Set([
    "moonc-version",
    "moonbit-release",
    "verified-date",
    "verified-platform",
    "verified-targets",
  ]),
  "moonbit-toolchain": new Set([
    "moon-version",
    "moonrun-version",
    "moonbit-release",
    "verified-date",
    "verified-platform",
    "verified-targets",
  ]),
};

export function validateSkill(skillDir: string): string[] {
  const problems: string[] = [];
  const skillName = basename(skillDir);
  const skillMd = join(skillDir, "SKILL.md");
  if (!isFile(skillMd)) {
    return [`${skillName}: missing SKILL.md`];
  }

  const text = readFileSync(skillMd, "utf8");
  const parsed = parseFrontmatter(text);
  problems.push(...parsed.errors.map((error) => `frontmatter: ${error}`));

  const name = scalar(parsed.frontmatter, "name") ?? "";
  if (!NAME_RE.test(name) || name.length > 64) {
    problems.push(`name ${repr(name)} violates spec naming rules`);
  }
  if (name !== skillName) {
    problems.push(`name ${repr(name)} != directory name ${repr(skillName)}`);
  }

  const description = scalar(parsed.frontmatter, "description") ?? "";
  if (description.length < 1 || description.length > 1024) {
    problems.push(`description length ${description.length} outside 1..1024 chars`);
  }

  const compatibility = scalar(parsed.frontmatter, "compatibility") ?? "";
  if (compatibility !== "" && compatibility.length > 500) {
    problems.push(`compatibility length ${compatibility.length} > 500 chars`);
  }

  let metadata = stringMap(parsed.frontmatter, "metadata");
  if (metadata === undefined) {
    problems.push("metadata must be a mapping");
    metadata = {};
  }
  const required = new Set([...REQUIRED_METADATA_KEYS, ...(REQUIRED_COMPONENT_KEYS[name] ?? [])]);
  const missing = [...required].filter((key) => !(key in metadata)).sort();
  if (missing.length > 0) {
    problems.push(`metadata missing version-contract keys: ${reprList(missing)}`);
  }
  const skillVersion = metadata["skill-version"];
  if (skillVersion !== undefined && !SKILL_VERSION_RE.test(skillVersion)) {
    problems.push(`metadata.skill-version ${repr(skillVersion)} must use SemVer x.y.z`);
  }
  const updatedDate = metadata["updated-date"];
  if (updatedDate !== undefined && !isIsoDate(updatedDate)) {
    problems.push(`metadata.updated-date ${repr(updatedDate)} must use a real YYYY-MM-DD date`);
  }
  const verifiedDate = metadata["verified-date"];
  if (verifiedDate !== undefined && !isIsoDate(verifiedDate)) {
    problems.push(`metadata.verified-date ${repr(verifiedDate)} must use a real YYYY-MM-DD date`);
  }
  if (name === "moonbit-agent-skills-maintainer" && metadata.scope !== "repository-maintenance") {
    problems.push("metadata.scope must be repository-maintenance");
  }
  if (metadata.scope === "repository-maintenance" && metadata.internal !== "true") {
    problems.push("repository-maintenance skills must set metadata.internal: true");
  }
  if (REQUIRED_COMPONENT_KEYS[name] !== undefined && "internal" in metadata) {
    problems.push("product skills must remain public (remove metadata.internal)");
  }

  const bodyLines = parsed.body.split("\n").length;
  if (bodyLines > 500) {
    problems.push(`SKILL.md body has ${bodyLines} lines (> 500)`);
  }
  const estimatedTokens = Math.floor(parsed.body.length / 4);
  if (estimatedTokens > 5000) {
    problems.push(`SKILL.md body ~${estimatedTokens} tokens (> 5000)`);
  }

  const referenced = new Set<string>();
  const pathPattern =
    /(?<![A-Za-z0-9_./-])((?:references|scripts|assets)\/[A-Za-z0-9_./-]+[A-Za-z0-9_])/g;
  for (const match of parsed.body.matchAll(pathPattern)) {
    referenced.add(match[1]);
  }
  for (const path of [...referenced].sort()) {
    if (!isFile(join(skillDir, path))) {
      problems.push(`SKILL.md references missing file ${path}`);
    }
  }

  for (const subdirectory of ["references", "scripts"]) {
    const directory = join(skillDir, subdirectory);
    if (!isDirectory(directory)) {
      continue;
    }
    for (const entry of readdirSync(directory, { withFileTypes: true }).sort((a, b) =>
      a.name.localeCompare(b.name),
    )) {
      if (entry.name.startsWith(".") || !entry.isFile()) {
        continue;
      }
      const path = `${subdirectory}/${entry.name}`;
      if (!referenced.has(path) && !parsed.body.includes(path)) {
        problems.push(`${path} exists but SKILL.md never mentions it`);
      }
    }
  }

  const sidecar = join(skillDir, "agents", "openai.yaml");
  if (!isFile(sidecar)) {
    problems.push("missing agents/openai.yaml (Codex activation sidecar)");
  } else if (!readFileSync(sidecar, "utf8").includes("allow_implicit_invocation: true")) {
    problems.push("agents/openai.yaml must set allow_implicit_invocation: true");
  }

  for (const file of walkFiles(skillDir)) {
    if (![".md", ".mbt", ".ts"].includes(extname(file))) {
      continue;
    }
    if (/(?<![\w@])\/(?:Users|home|private\/tmp)\//.test(readFileSync(file, "utf8"))) {
      problems.push(`${relative(skillDir, file)}: absolute path leaked`);
    }
  }

  return problems.map((problem) => `${skillName}: ${problem}`);
}

export function readmeCatalogProblems(readme: string, skillDirectories: string[]): string[] {
  const problems: string[] = [];
  const tableHeader = "| Skill | Version | Updated | MoonBit | Verified |";
  const sectionStart = readme.indexOf("## Skills\n");
  const sectionEnd = sectionStart === -1 ? -1 : readme.indexOf("\n## ", sectionStart + 10);
  const skillsSection =
    sectionStart === -1
      ? ""
      : readme.slice(sectionStart, sectionEnd === -1 ? readme.length : sectionEnd);
  const sectionLines = skillsSection.split("\n");
  const headerIndex = sectionLines.indexOf(tableHeader);
  if (headerIndex === -1) {
    problems.push(`README: missing public skill table header ${repr(tableHeader)}`);
  }
  const tableRows = new Set<string>();
  if (headerIndex !== -1) {
    for (const line of sectionLines.slice(headerIndex + 2)) {
      if (!line.startsWith("|")) break;
      tableRows.add(line);
    }
  }
  for (const skillDirectory of skillDirectories) {
    const skillMd = join(skillDirectory, "SKILL.md");
    if (!isFile(skillMd)) continue;
    const parsed = parseFrontmatter(readFileSync(skillMd, "utf8"));
    const name = scalar(parsed.frontmatter, "name") ?? basename(skillDirectory);
    if (REQUIRED_COMPONENT_KEYS[name] === undefined) continue;
    const metadata = stringMap(parsed.frontmatter, "metadata") ?? {};
    const status = readmeStatusRow(name, metadata);
    if (!tableRows.has(status)) {
      problems.push(`README: ${name} status does not match SKILL.md metadata`);
    }
  }
  return problems;
}

export function readmeStatusRow(name: string, metadata: Record<string, string>): string {
  return (
    `| [${name}](skills/${name}/SKILL.md) | ` +
    `[\`${metadata["skill-version"] ?? ""}\`](https://github.com/hyfdev/moonbit-agent-skills/commits/main/skills/${name}) | ` +
    `${metadata["updated-date"] ?? ""} | ` +
    `\`${metadata["moonbit-release"] ?? ""}\` | ` +
    `${metadata["verified-date"] ?? ""} |`
  );
}

export function isIsoDate(value: string): boolean {
  if (!DATE_RE.test(value)) return false;
  const date = new Date(`${value}T00:00:00Z`);
  return !Number.isNaN(date.valueOf()) && date.toISOString().slice(0, 10) === value;
}

export function main(): number {
  const skillDirectories = readdirSync(SKILLS_DIR, { withFileTypes: true })
    .filter((entry) => entry.isDirectory())
    .map((entry) => join(SKILLS_DIR, entry.name))
    .sort();
  if (skillDirectories.length === 0) {
    console.error("no skills found");
    return 2;
  }

  const problems = skillDirectories.flatMap(validateSkill);
  problems.push(
    ...readmeCatalogProblems(readFileSync(join(REPO_ROOT, "README.md"), "utf8"), skillDirectories),
  );
  for (const problem of problems) {
    console.error(`FAIL ${problem}`);
  }
  if (problems.length === 0) {
    console.log(`validated ${skillDirectories.length} skill(s): OK`);
  }
  return problems.length > 0 ? 1 : 0;
}

function isFile(path: string): boolean {
  return existsSync(path) && statSync(path).isFile();
}

function isDirectory(path: string): boolean {
  return existsSync(path) && statSync(path).isDirectory();
}

function walkFiles(directory: string): string[] {
  const files: string[] = [];
  for (const entry of readdirSync(directory, { withFileTypes: true })) {
    const path = join(directory, entry.name);
    if (entry.isDirectory()) {
      files.push(...walkFiles(path));
    } else if (entry.isFile()) {
      files.push(path);
    }
  }
  return files;
}

function repr(value: string): string {
  return `'${value.replaceAll("\\", "\\\\").replaceAll("'", "\\'")}'`;
}

function reprList(values: string[]): string {
  return `[${values.map(repr).join(", ")}]`;
}

if (isMain(import.meta.url)) {
  exitWith(main());
}
